testcases/crypto/dilithium_func.c

```diff
@@ -0,0 +1,297 @@
+    }
+
+    /* Verify */
+    rc = funcs->C_VerifyInit(session, &mech2, publ_key);
+    if (rc != CKR_OK) {
+        testcase_error("C_VerifyInit rc=%s", p11_get_ckr(rc));
+        goto testcase_cleanup;
+    }
+
+    rc = funcs->C_Verify(session, data, inputlen, signature, signaturelen);
+    if (rc != CKR_OK) {
+        testcase_error("C_Verify rc=%s", p11_get_ckr(rc));
+        goto testcase_cleanup;
+    }
+
+    /* Corrupt the signature and re-verify */
+    memcpy(signature, "ABCDEFGHIJKLMNOPQRSTUV",
+           strlen("ABCDEFGHIJKLMNOPQRSTUV"));
+
+    rc = funcs->C_VerifyInit(session, &mech2, publ_key);
+    if (rc != CKR_OK) {
+        testcase_error("C_VerifyInit rc=%s", p11_get_ckr(rc));
+        goto testcase_cleanup;
+    }
+
+    rc = funcs->C_Verify(session, data, inputlen, signature, signaturelen);
+    if (rc != CKR_SIGNATURE_INVALID) {
+        testcase_error("C_Verify rc=%s", p11_get_ckr(rc));
+        PRINT_ERR("		Expected CKR_SIGNATURE_INVALID\n");
+        goto testcase_cleanup;
+    }
+
+    rc = CKR_OK;
+
+testcase_cleanup:
+    if (data)
+        free(data);
+    if (signature)
+        free(signature);
+
+    return rc;
+}
+
+CK_RV run_GenerateDilithiumKeyPairSignVerify()
+{
+    CK_MECHANISM mech;
+    CK_OBJECT_HANDLE publ_key = CK_INVALID_HANDLE, priv_key = CK_INVALID_HANDLE;
+    CK_SESSION_HANDLE session;
+    CK_BYTE user_pin[PKCS11_MAX_PIN_LEN];
+    CK_ULONG user_pin_len, j;
+    CK_FLAGS flags;
+    CK_MECHANISM_INFO mech_info;
+    CK_RV rc;
+
+    testcase_begin("Starting Dilithium generate key pair.");
+
+    testcase_rw_session();
+    testcase_user_login();
+
+    mech.mechanism = CKM_IBM_DILITHIUM;
+    mech.ulParameterLen = 0;
+    mech.pParameter = NULL;
+
+    /* query the slot, check if this mech is supported */
+    rc = funcs->C_GetMechanismInfo(SLOT_ID, mech.mechanism, &mech_info);
+    if (rc != CKR_OK) {
+        if (rc == CKR_MECHANISM_INVALID) {
+            /* no support for EC key gen? skip */
+            testcase_skip("Slot %u doesn't support CKM_IBM_DILITHIUM ",
+                          (unsigned int) SLOT_ID);
+            rc = CKR_OK;
+            goto testcase_cleanup;
+        } else {
+            testcase_error("C_GetMechanismInfo() rc = %s", p11_get_ckr(rc));
+            goto testcase_cleanup;
+        }
+    }
+
+    /* Setup attributes for public/private Dilithium key */
+    CK_BBOOL attr_sign = TRUE;
+    CK_BBOOL attr_verify = TRUE;
+    CK_ATTRIBUTE dilithium_attr_private[] = {
+        {CKA_SIGN, &attr_sign, sizeof(CK_BBOOL)},
+    };
+    CK_ATTRIBUTE dilithium_attr_public[] = {
+        {CKA_VERIFY, &attr_verify, sizeof(CK_BBOOL)},
+    };
+
+    /* Generate Dilithium key pair */
+    rc = funcs->C_GenerateKeyPair(session, &mech,
+                   dilithium_attr_public, 1,
+                   dilithium_attr_private, 1,
+                   &publ_key, &priv_key);
+    testcase_new_assertion();
+    if (rc != CKR_OK) {
+        if (rc == CKR_MECHANISM_PARAM_INVALID ||
```

Comment:
Why this check? Shouldnt the above C_getMechanismInfo have checked if that mech is supported?

Reply:
Hmm ... this check does in fact more than necessary with our current Dilithium support: Dilithium does not support a mechanism parameter, so CKR_MECHANISM_PARAM_INVALID cannot be returned. But we also don't have special attributes, like EC with different curves, where some curves are not supported. I therefore removed this check completely. Let me know if you think we should be more defensive here ...